lib/std/meta.zig

```diff
@@ -1129,15 +1129,78 @@ pub inline fn hasFn(comptime T: type, comptime name: []const u8) bool {
     return @typeInfo(@TypeOf(@field(T, name))) == .Fn;
 }
 
+test "hasFn" {
+    const S1 = struct {
+        pub fn foo() void {}
+    };
+
+    try std.testing.expect(hasFn(S1, "foo"));
+    try std.testing.expect(!hasFn(S1, "bar"));
+    try std.testing.expect(!hasFn(*S1, "foo"));
+
+    const S2 = struct {
+        foo: fn () void,
+    };
+
+    try std.testing.expect(!hasFn(S2, "foo"));
+
+    const S3 = struct {
+        fn foo() void {}
+    };
+
+    try std.testing.expect(!hasFn(S3, "foo"));
+}
+
 /// Returns true if a type has a `name` method; `false` otherwise.
 /// Result is always comptime-known.
 pub inline fn hasMethod(comptime T: type, comptime name: []const u8) bool {
     return switch (@typeInfo(T)) {
-        .Pointer => |P| hasFn(P.child, name),
+        .Pointer => |P| switch (P.size) {
+            .One => hasFn(P.child, name),
+            .Many, .Slice, .C => false,
+        },
         else => hasFn(T, name),
     };
 }
 
+test "hasMethod" {
+    try std.testing.expect(!hasMethod(u32, "foo"));
+    try std.testing.expect(!hasMethod([]u32, "len"));
+    try std.testing.expect(!hasMethod(struct { u32, u64 }, "len"));
+
+    const S1 = struct {
+        pub fn foo() void {}
+    };
+
+    try std.testing.expect(hasMethod(S1, "foo"));
+    try std.testing.expect(hasMethod(*S1, "foo"));
```

Comment:
oh this function is much more flawed than i first realized, though it is the inner part of the actual correct algorithm.

the first should only pass if `foo` was `pub fn foo(_: S1) void {}` or `pub fn foo(_: S2) void {}`
the second should never pass.

Reply:
pretty sure this'll do it

```zig
pub inline fn hasMethod(comptime T: type, comptime name: []const u8) ?bool {
    return comptime switch (@typeInfo(T)) {
        .Struct, .Union, .Enum => blk: {
            if (!hasFn(T, name)) break :blk false;
            const params = @typeInfo(@TypeOf(@field(T, name))).Fn.params;
            if (params.len == 0) break :blk false;
            const First = params[0].type orelse break :blk null;
            break :blk switch (@typeInfo(First)) {
                .Pointer => |P| switch (P.size) {
                    .One => hasFn(P.child, name),
                    .Many, .Slice, .C => false,
                },
                .Struct, .Enum, .Union => First == T,
                else => false,
            };
        },
        else => false,
    };
}
```

Reply:
If I'm not mistaken, that second test case with `*S1` was the main reason the `hasMethod` function was introduced.

Reply:
Would it be okay to keep the function as it is, but just give it a better name? (something like `canCallFunction`....)